crates/storage/provider/src/providers/snapshot/manager.rs

```diff
@@ -71,23 +201,70 @@ impl SnapshotProvider {
             .as_ref()
             .and_then(|tracker| tracker.borrow().and_then(|highest| highest.highest(segment)))
     }
+
+    /// Iterates through segment snapshots in reverse order, executing a function until it returns
+    /// some object. Useful for finding objects by [`TxHash`] or [`BlockHash`].
+    pub fn find_snapshot<T>(
+        &self,
+        segment: SnapshotSegment,
+        func: impl Fn(SnapshotJarProvider<'_>) -> RethResult<Option<T>>,
+    ) -> RethResult<Option<T>> {
+        let snapshots = self.snapshots_block_index.read();
+        if let Some(segment_snapshots) = snapshots.get(&segment) {
+            // It's more probable that the request comes from a newer block height, so we iterate
+            // the snapshots in reverse.
+            let mut snapshots_rev_iter = segment_snapshots.iter().rev().peekable();
+
+            while let Some((block_end, tx_range)) = snapshots_rev_iter.next() {
+                let block_start =
+                    snapshots_rev_iter.peek().map(|(block_end, _)| *block_end + 1).unwrap_or(0);
+
+                if let Some(res) = func(self.get_or_create_jar_provider(
+                    segment,
+                    &(block_start..=*block_end),
+                    tx_range,
+                )?)? {
+                    return Ok(Some(res))
+                }
+            }
+        }
+
+        Ok(None)
+    }
 }
 
 impl HeaderProvider for SnapshotProvider {
-    fn header(&self, _block_hash: &BlockHash) -> RethResult<Option<Header>> {
-        todo!()
+    fn header(&self, block_hash: &BlockHash) -> RethResult<Option<Header>> {
+        self.find_snapshot(SnapshotSegment::Headers, |jar_provider| {
+            Ok(jar_provider
+                .cursor()?
+                .get_two::<HeaderMask<Header, BlockHash>>(block_hash.into())?
+                .and_then(|(header, hash)| {
+                    if &hash == block_hash {
+                        return Some(header)
+                    }
+                    None
+                }))
+        })
     }
```

Comment:
this is very slow, right? Because in the worst case of non-existing block hash, we iterate all rows in all snapshots. Is this a DoS vector? Maybe we should separate the `HeaderProvider` into methods that can be queried efficiently with snapshots, and those that are not?

Reply:
we never iterate all rows here, only all snapshots.

Ideally (still up for decision), the `DatabaseProvider` which is the one calling the snapshotprovider, will first check its index (hash -> number), and actually query `snapshot_provider.header_by_num(_)`.

However, i think it's still valuable to have this possibility, even if not directly used in the RPC

Reply:
if it's the latest (or close) snapshot, it should actually be faster from the tests i've done

Reply:
underneath it's using the perfect hashing functions, and not actually iterating all rows

Reply:
> underneath it's using the perfect hashing functions

ah yes, right. `jar_provider` contains snapshots for the specified segments, and we check each snapshot with one call. SGTM.